addon/components/base/bs-form.js

```diff
@@ -343,64 +343,67 @@ export default Component.extend({
       return;
     }
 
-    this.set('isSubmitting', true);
-    this.incrementProperty('pendingSubmissions');
-
     let model = this.get('model');
+    let record;
 
+    this.incrementProperty('pendingSubmissions');
     this.get('onBefore')(model);
 
-    RSVP.resolve(this.get('hasValidator') ? this.validate(this.get('model')) : null)
-      .then(
-        (record) => {
-          return RSVP.resolve()
-            .then(() => {
-              if (this.get('hideValidationsOnSubmit') === true) {
-                this.set('showAllValidations', false);
-              }
-              return this.get('onSubmit')(model, record);
-            })
-            .then(
-              () => {
-                if (!this.get('isDestroyed')) {
-                  this.set('isSubmitted', true);
-                }
-              },
-              () => {
-                if (!this.get('isDestroyed')) {
-                  this.set('isRejected', true);
-                }
-              }
-            );
-        },
-        (error) => {
-          this.set('showAllValidations', true);
-
-          return RSVP.resolve()
-            .then(() => {
-              return this.get('onInvalid')(model, error);
-            })
-            .finally(() => {
-              if (!this.get('isDestroyed')) {
-                this.set('isRejected', true);
-              }
-            });
-        }
-      )
-      .finally(() => {
-        if (!this.get('isDestroyed')) {
-          if (this.get('pendingSubmissions') === 1) {
-            this.set('isSubmitting', false);
-          }
-
-          this.decrementProperty('pendingSubmissions');
-
-          // reset forced hiding of validations
-          if (this.get('showAllValidations') === false) {
-            this.set('showAllValidations', undefined);
-          }
-        }
-      });
+    try {
+      record = await (this.get('hasValidator') ? this.validate(model) : null);
+    } catch (error) {
+      // model is invalid
+      this.set('showAllValidations', true);
+
+      await this.get('onInvalid')(model, error);
+
+      if (this.get('isDestroyed')) {
+        return;
+      }
+
+      this.set('isRejected', true);
+      this.decrementProperty('pendingSubmissions');
+
+      throw error;
```

Comment:
Should we reject the validation error or a generic error? Do we consider the value the Promise resolves with as part of public API or just the promise state (rejected / fulfilled)?

Reply:
> Should we reject the validation error or a generic error

The former I would say. So keep it like it is.

> Do we consider the value the Promise resolves with as part of public API or just the promise state

Hm, good question. I would say also the former!?